codec/dagjson/multicodec.go

```diff
@@ -17,13 +17,23 @@ var (
 
 func init() {
 	multicodec.RegisterEncoder(0x0129, Encode)
+	multicodec.RegisterEncoder(0x0200, Encode)
```

Comment:
I'm inclined to think that for the json codec, if we encounter a Link in the middle of a structure, we should have the codec error explicitly.  It's not going to roundtrip via that same codec, and explicit errors are better than unstated incorrect behavior.

Reply:
it produces a reasonable serialization into json of a dag. making the caller traverse the dag to understand there's no links in it, rather than just flatting them isn't clear to me as a win

Reply:
I don't think callers generally have to _do_ anything.  They would just... get an error if they try to encode data that's unencodable with this format.

In general, things that are unencodable should result in errors.  That's a rule we'd follow anywhere[‡] that isn't JSON, and I think we should follow it for JSON too.

Failure to have bijections for data that we emit tends to cause unhappiness and confusion.  The problems are magnified by the fact they often sweep under the rug _for a while_ and then often reappear at points in time distant from when the data was first written.

I'd say the direction of burden-of-proof should be overwhelming weighted such that doing something looseygoosey that's non-bijective is the case that needs support.  Maintaining bijectiveness, even if it comes at the cost of more errors, should be the default design bias.

[‡] - except where floating point numbers are involved.  But that's the exception that really proves the reason for the rule, isn't it -- float consistency is a horrific minefield that we would deeply prefer didn't exist, and there are horrific amounts of time wasted by this seemingly small problem.

Reply:
@warpfork my understanding was that people should be able to take DAG-JSON data and by changing the codec to JSON they'd just have an object where there are no links. This is similar to how we can take an existing CID and slap a Raw codec on it and get the data in that block back as raw bytes.

This case is mostly covered by the decode case, it seems a bit weird that I wouldn't be able to roundtrip data since Encode(Decode(dagJSONDataAsJSON)) would error. Am I missing something?

Reply:
when you decode as json you don't  get a link. when you re-encode the map with a string in it you'd get the same data.

Reply:
maybe we're flipping around `decode` and `encode`?  decoding will realize some ipld-prime nodes. you'll end up with a map containing a string.
When you go to re-encode it to serialized json it'll work fine. it's when you encode an IPLD prime link node that it would get unhappy with you.

Reply:
As a general rule, there are _no_ guarantees about what happens when you take some data and try to wangle it with a _different_ codec.  Maybe it works out fine.  Maybe it doesn't.  It probably doesn't.  In the cases it does, that was very "lucky" and probably based mostly on coincidences.

I'm concerned with bijection *within the same codec*.

Reply:
`{"/":"bafyquux"}` is a map with one key in JSON.  When you deserialize that block with a JSON codec, you get a map with one key; when you serialize it with a JSON codec, you get the same block.

(@aschmahmann , the reason I thumbs-down'd your comment is that there's no error in this case above.  One just emits this as a map.  That's normal.)

`{"/":"bafyquux"}` is a link in DAG-JSON.  When you deserialize that block with a DAG-JSON codec, you get a link; when you serialize it with a DAG-JSON codec, you get the same block.

Now, I have a link, and I attempt serialize it with a JSON codec, I want that to yield an error, because that's not really encodable in JSON.  As a user, if I attempt to do this, _I have made a serious mistake_, and I _want to be informed immediately_.

If we didn't do this, and instead smash the link into a map with one entry during the encode process, we've changed the data.  When we deserialize it later with the *same* codec we serialized it with, we would get different data.  This would not be good.

I don't really care what happens when we serialize something with a JSON codec and attempt to later deserialize it with a DAG-JSON codec.  If that gives different data... well, yes.  A different codec was used.  That's gonna happen.

Reply:
It may help to be reminded that neither DAG-JSON nor JSON are total supersets of each other.  This sucks, and we often wish it wasn't so, but wishes... well.   Wishes.  They have limited power.

DAG-JSON can't encode maps with single entries that happen to have the key "/".  JSON can.  This makes JSON larger.

In the other direction, DAG-JSON can encode links.  JSON can't.  This makes DAG-JSON larger.

This is generally a headache.  But I'm fairly certain we make the headache _bigger_ rather than smaller if we were to add more silent coercions to the picture.

Reply:
np, just wanted to clarify what you were actually intending/trying to do. I have no problem with a codec's `Encode` function erroring when trying to convert an IPLD data model object to bytes when that codec doesn't support part of the data model used in that object (e.g. JSON doesn't support links).

Reply:
> DAG-JSON can't encode maps with single entries that happen to have the key "/".

@warpfork maybe I'm missing it, but it doesn't look like DAG-JSON's `Encode` function errors with these map entries. Should that be fixed in another PR?

Reply:
... :weary: yes, you're quite correct